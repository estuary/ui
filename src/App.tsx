import { Skeleton } from '@mui/material';
import AppLayout from 'AppLayout';
import RequireAuth from 'auth/RequireAuth';
import Home from 'pages/Home';
import Login from 'pages/Login';
import LoginHelp from 'pages/LoginHelp';
import React, { Suspense } from 'react';
import { Route, Routes } from 'react-router';
import Error from './pages/Error';

const Admin = React.lazy(() => import('./pages/Admin'));

const Capture = React.lazy(() => import('./pages/Captures'));
const NewCaptureModal = React.lazy(
    () => import('components/capture/creation/index')
);

const Collections = React.lazy(() => import('./pages/Collections'));

const Users = React.lazy(() => import('./pages/Users'));

const Alerts = React.lazy(() => import('./pages/Alerts'));

const Logs = React.lazy(() => import('./pages/Logs'));

const App: React.FC = () => {
    return (
        <Suspense fallback={<Skeleton animation="wave" />}>
            <Routes>
                {
                    // TODO - do we need the default path to use a "redirector"/"landing" page kind of thing?
                }
                <Route path="" element={<Login />} />
                <Route path="login/*">
                    <Route path="" element={<Login />} />
                    <Route path="help" element={<LoginHelp />} />
                </Route>

                <Route
                    element={
                        <RequireAuth>
                            <AppLayout />
                        </RequireAuth>
                    }
                >
                    <Route path="/dashboard" element={<Home />} />
                    <Route path="/app">
                        <Route path="collections" element={<Collections />} />
                        <Route path="captures" element={<Capture />}>
                            <Route path="new" element={<NewCaptureModal />} />
                        </Route>
                        <Route path="derivations" element={<Error />} />
                        <Route path="materializations" element={<Error />} />
                        <Route path="admin/*" element={<Admin />}>
                            <Route path="logs" element={<Logs />} />
                            <Route path="alerts" element={<Alerts />} />
                            <Route path="users" element={<Users />} />
                        </Route>
                    </Route>
                </Route>
                <Route path="*" element={<Error />} />
            </Routes>
        </Suspense>
    );
};

export default App;
