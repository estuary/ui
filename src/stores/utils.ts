import { isEmpty, map } from 'lodash';

export const checkForErrors = (obj: any) =>
    obj?.errors && obj.errors.length > 0;

export const populateErrors = (configDictionary: any) => {
    let configErrors: any[] = [];

    // TODO (errors) When there are not configs do we need to populate this object with something?
    if (Object.keys(configDictionary).length > 0) {
        map(configDictionary, (config) => {
            if (checkForErrors(config)) {
                configErrors = configErrors.concat(config.errors);
            }
        });
    }

    return {
        configErrors,
        hasErrors: !isEmpty(configErrors),
    };
};
