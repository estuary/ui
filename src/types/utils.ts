export type WithRequiredProperty<T, K extends keyof T> = T & {
    [P in K]-?: T[P];
};

export type WithRequiredNonNullProperty<T, K extends keyof T> = T & {
    [P in K]-?: Exclude<T[P], null>;
};
