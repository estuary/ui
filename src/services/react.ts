import FullPageSpinner from 'components/fullPage/Spinner';
import { lazy } from 'react';
import { logRocketConsole, logRocketEvent } from './shared';
import { CustomEvents } from './types';

const LAZY_LOAD_FAILED_KEY = 'chunk_failed';

// https://mitchgavan.com/code-splitting-react-safely/
const setWithExpiry = (key: string, value: string, ttl: number) => {
    const item = {
        value,
        expiry: new Date().getTime() + ttl,
    };
    localStorage.setItem(key, JSON.stringify(item));
};

const getWithExpiry = (key: string) => {
    const itemString = window.localStorage.getItem(key);
    const item = itemString ? JSON.parse(itemString) : null;

    if (item === null) {
        return null;
    }

    // We have waited long enough to allow trying again so clearing out the key
    if (new Date().getTime() > item.expiry) {
        localStorage.removeItem(key);
        return null;
    }

    // Return value so we do not try reloading again
    return item.value;
};

const handledLazy = (factory: () => Promise<{ default: any }>) => {
    return lazy(() =>
        factory().catch((error) => {
            logRocketEvent(CustomEvents.LAZY_LOADING, 'failed');
            logRocketConsole('Component Failed Loading 1:', error);

            // Check if we're in an infinite loading loop before trying again
            if (!getWithExpiry(LAZY_LOAD_FAILED_KEY)) {
                setWithExpiry(LAZY_LOAD_FAILED_KEY, 'true', 10000);
                window.location.reload();
            }

            // Still make sure to return something so the app does not blow up
            return { default: FullPageSpinner };
        })
    );
};

export { handledLazy };
