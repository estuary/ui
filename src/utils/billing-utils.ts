import { isEqual, parseISO } from 'date-fns';
import { isEmpty, sum } from 'lodash';
import prettyBytes from 'pretty-bytes';
import { BillingRecord } from 'stores/Billing/types';
import { CatalogStats_Billing, Entity } from 'types';

export const TOTAL_CARD_HEIGHT = 300;

// Grid item height - 72 = graph canvas height
export const CARD_AREA_HEIGHT = TOTAL_CARD_HEIGHT - 72;

export const BYTES_PER_GB = 1073741824;

const FREE_TASK_COUNT = 2;

export enum FREE_GB_BY_TIER {
    FREE = 20,
    PERSONAL = 10,
    ENTERPRISE = 10,
}

export const evaluateSpecType = (query: CatalogStats_Billing): Entity => {
    if (Object.hasOwn(query.flow_document.taskStats, 'capture')) {
        return 'capture';
    } else if (Object.hasOwn(query.flow_document.taskStats, 'materialize')) {
        return 'materialization';
    } else {
        return 'collection';
    }
};

const evaluateTotalCost = (dataVolume: number, taskCount: number) => {
    const freeBytes = 10 * BYTES_PER_GB;

    const dataVolumeOverLimit =
        dataVolume > freeBytes ? dataVolume - freeBytes : 0;

    const taskCountOverLimit =
        taskCount > FREE_TASK_COUNT ? taskCount - FREE_TASK_COUNT : 0;

    return (
        (dataVolumeOverLimit / BYTES_PER_GB) * 0.75 + taskCountOverLimit * 20
    );
};

const evaluateDataVolume = (stats: CatalogStats_Billing[]): number => {
    const taskBytes: number[] = stats.map((query) => {
        if (Object.hasOwn(query.flow_document.taskStats, 'capture')) {
            return query.bytes_written_by_me;
        } else if (
            Object.hasOwn(query.flow_document.taskStats, 'materialize')
        ) {
            return query.bytes_read_by_me;
        } else {
            return query.bytes_written_by_me + query.bytes_read_by_me;
        }
    });

    return sum(taskBytes);
};

export const stripTimeFromDate = (date: string) => {
    const [truncatedDateStr] = date.split('T', 1);

    return parseISO(truncatedDateStr);
};

const getInitialBillingRecord = (date: string): BillingRecord => {
    const truncatedDate = stripTimeFromDate(date);

    return {
        date: truncatedDate,
        dataVolume: 0,
        taskCount: 0,
        totalCost: 0,
        pricingTier: null,
        taskRate: null,
        gbFree: null,
        includedTasks: null,
    };
};

// TODO (billing): Remove this helper function to translate data returned from
//   the new RPC when available.
export const formatBillingCatalogStats = (
    value: CatalogStats_Billing[]
): BillingRecord[] => {
    const taskStatData = value.filter((query) =>
        Object.hasOwn(query.flow_document, 'taskStats')
    );

    let sortedStats: {
        [date: string]: CatalogStats_Billing[];
    } = {};

    taskStatData.forEach((query) => {
        if (Object.hasOwn(sortedStats, query.ts)) {
            sortedStats[query.ts].push(query);
        } else {
            sortedStats = {
                ...sortedStats,
                [query.ts]: [query],
            };
        }
    });

    const billingHistory: BillingRecord[] = [];

    if (!isEmpty(sortedStats)) {
        Object.entries(sortedStats).forEach(([ts, stats]) => {
            const billingRecordIndex = billingHistory.findIndex((record) =>
                isEqual(record.date, stripTimeFromDate(ts))
            );

            const taskCount = stats.length;
            const dataVolume = evaluateDataVolume(stats);
            const totalCost = evaluateTotalCost(dataVolume, taskCount);

            if (billingRecordIndex === -1) {
                const { date, pricingTier, taskRate, gbFree, includedTasks } =
                    getInitialBillingRecord(ts);

                billingHistory.push({
                    date,
                    dataVolume,
                    taskCount,
                    totalCost,
                    pricingTier: pricingTier ?? 'personal',
                    taskRate: taskRate ?? 20,
                    gbFree: gbFree ?? FREE_GB_BY_TIER.PERSONAL,
                    includedTasks: includedTasks ?? 2,
                });
            } else {
                const { date, pricingTier, taskRate, gbFree, includedTasks } =
                    billingHistory[billingRecordIndex];

                billingHistory[billingRecordIndex] = {
                    date,
                    dataVolume,
                    taskCount,
                    totalCost,
                    pricingTier: pricingTier ?? 'personal',
                    taskRate: taskRate ?? 20,
                    gbFree: gbFree ?? FREE_GB_BY_TIER.PERSONAL,
                    includedTasks: includedTasks ?? 2,
                };
            }
        });
    }

    return billingHistory;
};

export interface SeriesConfig {
    data: [string, number][];
    seriesName?: string;
    stack?: string;
}

export const formatDataVolumeForDisplay = (
    seriesConfigs: SeriesConfig[],
    tooltipConfig: any
): string => {
    const filteredSeries =
        seriesConfigs.length === 1
            ? seriesConfigs
            : seriesConfigs.filter(
                  (series) => series.seriesName === tooltipConfig.seriesName
              );

    const dataVolumeInBytes = filteredSeries
        .flatMap(({ data }) => data)
        .find(([month]) => month === tooltipConfig.name);

    return dataVolumeInBytes
        ? prettyBytes(dataVolumeInBytes[1], { minimumFractionDigits: 2 })
        : `${tooltipConfig.value[1]} GB`;
};
