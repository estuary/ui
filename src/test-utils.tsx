// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render, RenderOptions } from '@testing-library/react';
import { ReactElement, type FC } from 'react';
import { act } from 'react-dom/test-utils';
import { localStorageKey } from './auth';
import AppProviders from './context';

const AllTheProviders: FC = ({ children }) => {
    return <AppProviders>{children}</AppProviders>;
};

const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

const loginAsUser = (userName: string = 'fakeUserName') => {
    act(() => {
        window.localStorage.setItem(localStorageKey, userName);
    });
};

export * from '@testing-library/react';
export { customRender as render, loginAsUser };
