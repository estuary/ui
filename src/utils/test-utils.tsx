/* eslint-disable @typescript-eslint/unbound-method */
// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { createClient, User } from '@supabase/supabase-js';
import { Auth } from '@supabase/ui';
import { AuthSession } from '@supabase/ui/dist/cjs/components/Auth/UserContext';
import { render as rtlRender, RenderOptions } from '@testing-library/react';
import AppContent from 'context/Content';
import AppRouter from 'context/Router';
import ThemeProvider from 'context/Theme';
import produce from 'immer';
import { mockDeep } from 'jest-mock-extended';
import { ReactElement } from 'react';
import { SwrSupabaseContext } from 'supabase-swr';

const goTo = (route?: string, name?: string) => {
    window.history.pushState(
        {},
        name ? name : 'Test page',
        route ? route : '/'
    );
};

const waitForLoading = async () => {
    return Promise.resolve();
};

// TODO - We use immer here but don't need it. The user metadata is currently just
//  any number of keys (according to the types). Hoping that changes in the future.
const generateMockUserMetadata = (username: string) => {
    return produce(mockDeep<User['user_metadata']>(), (draft) => {
        draft.avatar_url = `https://example.org/avatar/${username}`;
        draft.email = `${username}@example.org`;
        draft.email_verified = true;
        draft.full_name = `Full ${username}`;
        draft.iss = 'https://api.example.org';
        draft.name = username;
        draft.picture = `https://example.org/picture/${username}`;
        draft.preferred_username = username;
        draft.provider_id = '102885624013818367764';
        draft.sub = '102885624013818367764';
        draft.user_name = username;

        return draft;
    });
};

const MockProviders = ({ children, username }: any) => {
    const mockClient = createClient('http://mock.url', 'MockSupabaseAnonKey');
    const mockAuthSession = mockDeep<AuthSession>();

    if (username) {
        mockAuthSession.user = mockDeep<User>();
        mockAuthSession.user.user_metadata = generateMockUserMetadata(username);
        Auth.useUser = () => mockAuthSession;
    }

    return (
        <SwrSupabaseContext.Provider value={mockClient}>
            <Auth.UserContextProvider supabaseClient={mockClient}>
                {children}
            </Auth.UserContextProvider>
        </SwrSupabaseContext.Provider>
    );
};

const customRender = async (
    ui: ReactElement,
    options: Omit<RenderOptions, 'wrapper'> & {
        route?: string;
        username?: string;
    }
) => {
    const { route, username } = options;

    goTo(route, 'Test Page');

    const view = rtlRender(
        <AppContent>
            <MockProviders username={username}>
                <ThemeProvider>
                    <AppRouter>{ui}</AppRouter>
                </ThemeProvider>
            </MockProviders>
        </AppContent>,
        {
            ...options,
        }
    );

    await waitForLoading();

    return view;
};

export * from '@testing-library/react';
export { generateMockUserMetadata, goTo, customRender, waitForLoading };
