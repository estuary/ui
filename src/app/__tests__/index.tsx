import '@testing-library/jest-dom';
import { screen, waitFor } from '@testing-library/react';
import { customRender } from 'utils/test-utils';
import App from '..';

test('Unauthenticated app renders when no user is present', async () => {
    await customRender(<App />, {
        username: undefined,
    });

    await waitFor(() => {
        expect(
            screen.getByText(
                'Please use one of the providers below to continue.'
            )
        ).toBeInTheDocument();
    });
});
