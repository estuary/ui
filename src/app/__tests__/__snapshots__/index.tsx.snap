// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When there is a user the authorized app is logged into 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-1ji7omv-MuiBackdrop-root"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      >
        <span
          aria-label="Loading..."
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorInherit css-62e83j-MuiCircularProgress-root"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-1ji7omv-MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <span
        aria-label="Loading..."
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorInherit css-62e83j-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`When there is no user the login page is displayed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-1ji7omv-MuiBackdrop-root"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      >
        <span
          aria-label="Loading..."
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorInherit css-62e83j-MuiCircularProgress-root"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-1ji7omv-MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <span
        aria-label="Loading..."
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorInherit css-62e83j-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
